apps/gateway/test/champions_test.exs

```diff
@@ -0,0 +1,286 @@
+defmodule Gateway.Test.Champions do
+  @moduledoc """
+  Test for Champions of Mirra messages.
+  """
+  use ExUnit.Case
+
+  alias GameBackend.Users.Currencies.CurrencyCost
+  alias Champions.{Units, Users, Utils}
+  alias GameBackend.Users.Currencies
+
+  alias Gateway.Serialization.{
+    Currency,
+    Error,
+    Unit,
+    UnitAndCurrencies,
+    User,
+    UserCurrency,
+    WebSocketResponse
+  }
+
+  alias Gateway.SocketTester
+
+  # import Plug.Conn
+  # import Phoenix.ConnTest
+
+  setup_all do
+    # Start Phoenix endpoint
+    {:ok, _} =
+      Plug.Cowboy.http(Gateway.Endpoint, [],
+        ip: {127, 0, 0, 1},
+        port: 4001,
+        dispatch: [
+          _: [{"/2", Gateway.ChampionsSocketHandler, []}]
+        ]
+      )
+
+    :ok
+  end
+
+  setup do
+    {:ok, socket_tester} = SocketTester.start_link()
+
+    {:ok, %{socket_tester: socket_tester}}
+  end
+
+  describe "users" do
+    test "users", %{socket_tester: socket_tester} do
+      # Create our user
+      :ok = SocketTester.create_user(socket_tester, "Username")
+      fetch_last_message(socket_tester)
+      assert_receive %WebSocketResponse{response_type: {:user, %User{}}}
+
+      fetch_last_message(socket_tester)
+      %WebSocketResponse{response_type: {:user, user}} = get_last_message()
+
+      # Creating another user with the same name fails
+      :ok = SocketTester.create_user(socket_tester, "Username")
+      fetch_last_message(socket_tester)
+      assert_receive %WebSocketResponse{response_type: {:error, %Error{reason: "username_taken"}}}
+
+      # Get user by name
+      :ok = SocketTester.get_user_by_username(socket_tester, "Username")
+      fetch_last_message(socket_tester)
+      assert_receive %WebSocketResponse{response_type: {:user, ^user}}
+    end
+  end
+
+  describe "units" do
+    test "unit selection", %{socket_tester: socket_tester} do
+      {:ok, user} = Users.register("SelectUser")
+
+      [unit_to_unselect | _] = user.units
+      slot = unit_to_unselect.slot
+
+      # Unit is selected by default (this will change when we remove sample data in user creation)
+      assert unit_to_unselect.selected
+
+      # Unselect the unit
+      :ok = SocketTester.unselect_unit(socket_tester, user.id, unit_to_unselect.id)
+      fetch_last_message(socket_tester)
+      assert_receive %WebSocketResponse{response_type: {:unit, %Unit{}}}
+
+      fetch_last_message(socket_tester)
+      %WebSocketResponse{response_type: {:unit, unselected_unit}} = get_last_message()
```

Comment:
Maybe we should also check that the `unselected_unit` id is the same as the `unit_to_unselect` to ensure the unit is the same

Reply:
Done!